framework/src/media/audio/audio_manager.c

```diff
@@ -600,52 +658,46 @@ int start_audio_stream_out(void *data, uint32_t frames)
 		medvdbg("Resume the output audio card!!\n");
 	}
 
+	if (cur_card->resample.necessary) {
+		resampled_frames = resample_stream_out(pcm, cur_card, data, frames);
+	}
+
 	do {
-		medvdbg("Start Playing!! Resample : %d\n", cur_card->resample.necessary);
+		medvdbg("Start Playing!! Resample : %d\t", cur_card->resample.necessary);
 
-		// Todo: Need to consider xrun situation.
 		if (cur_card->resample.necessary) {
-			int frames_total = frames;
-			int frames_used = 0;
-			src_data_t srcData = { 0, };
-
-			while (frames_total > frames_used) {
-				srcData.data_in = (void *)data + get_output_frames_byte_size(frames_used);
-				srcData.input_frames = frames_total - frames_used;
-				srcData.channels_num = 2;
-				srcData.origin_sample_rate = cur_card->resample.from;
-				srcData.origin_sample_width = SAMPLE_WIDTH_16BITS;
-				srcData.desired_sample_rate = cur_card->resample.to;
-				srcData.desired_sample_width = SAMPLE_WIDTH_16BITS;
-				srcData.input_frames_used = 0;
-
-				if (src_simple(cur_card->resample.handle, &srcData) != SRC_ERR_NO_ERROR) {
-					meddbg("Fail to resample to %u from %u\n", srcData.desired_sample_rate, srcData.origin_sample_rate);
-					return AUDIO_MANAGER_RESAMPLE_FAIL;
-				}
-
-				frames_used += srcData.input_frames_used;
-
-				ret = pcm_writei(pcm, srcData.data_out, srcData.output_frames_gen);
-				medvdbg("pcm_writei: %d/%d Resampled(%d/%d)\n", ret, srcData.output_frames_gen, frames_used, frames_total);
-			}
-		} else {
+			ret = pcm_writei(pcm, cur_card->resample.buffer, resampled_frames);
+		}
+		else {
 			ret = pcm_writei(pcm, data, frames);
-			medvdbg("pcm_writei return: %d [%02x%02x%02x%02x]\n", ret, ((unsigned char *)data)[0], ((unsigned char *)data)[1], ((unsigned char *)data)[2], ((unsigned char *)data)[3]);
 		}
 
 		if (ret == -EPIPE) {
-			ret = pcm_prepare(pcm);
-			if (ret != OK) {
-				meddbg("Fail to pcm_prepare()\n");
+			if (prepare_retry > 0) {
+				ret = pcm_prepare(pcm);
+				if (ret != OK) {
+					meddbg("Fail to pcm_prepare()\n");
+					return AUDIO_MANAGER_XRUN_STATE;
+				}
+				prepare_retry--;
+			}
+			else {
+				meddbg("prepare_retry = 0\n");
 				return AUDIO_MANAGER_XRUN_STATE;
 			}
-		} else if (ret == -EINVAL) {
+		}
+		else if (ret == -EINVAL) {
+			meddbg("pcm_writei = -EINVAL\n");
 			return AUDIO_MANAGER_INVALID_PARAM;
-		} else {
+		}
+		else {
+			if (ret > 0) {
+				ret = frames;
```

Comment:
wrong logic here, we should return ret, then above layer can check how many frames processed.